tests/CodeBehind/UnitTests/BuildTests.cs

```diff
@@ -594,35 +619,43 @@ string Regexify (string input)
 				.Replace ("}", "\\}");
 		}
 
-		[OneTimeSetUp]
-		public void BuildTestsSetUp ()
+		string PrepareProject (string testName)
 		{
-			if (!File.Exists (TestProjectPath))
-				throw new InvalidOperationException ($"Test project '{TestProjectPath}' not found");
+			string tempRoot = Path.Combine (TestOutputDir, $"{testName}.build", XABuildPaths.Configuration);
+			string temporaryProjectPath = Path.Combine (tempRoot, "project");
+			CopyRecursively (TestProjectRootDirectory, temporaryProjectPath);
+			CopyRecursively (CommonSampleLibraryRootDirectory, Path.Combine (tempRoot, CommonSampleLibraryName));
+			return temporaryProjectPath;
 		}
 
-		void ResetProject (string testName)
+		void CopyRecursively (string fromDir, string toDir)
 		{
-			RemoveDir (TestProjectObjPath);
-			RemoveDir (TestProjectBinPath);
+			if (String.IsNullOrEmpty (fromDir))
+				throw new ArgumentException ($"{nameof (fromDir)} is must have a non-empty value");
+			if (String.IsNullOrEmpty (toDir))
+				throw new ArgumentException ($"{nameof (toDir)} is must have a non-empty value");
 
-			string outputDir = GetTestOutputDir (testName);
-			if (Directory.Exists (outputDir))
-				Directory.Delete (outputDir, true);
+			var fdi = new DirectoryInfo (fromDir);
+			if (!fdi.Exists)
+				throw new InvalidOperationException ($"Source directory '{fromDir}' does not exist");
 
-			RemoveLogs (testName);
-		}
+			if (Directory.Exists (toDir))
```

Comment:
You don't need to wrap `Directory.Delete()` with `Directory.Exists()`; `Directory.Delete()` is a no-op if the directory doesn't exist.

Reply:
Not quite:

```
csharp> Directory.Delete("/tmp/tes")          
csharp> Directory.Delete("/tmp/tes")         
System.IO.DirectoryNotFoundException: Directory does not exist.
  at System.IO.Directory.Delete (System.String path) [0x0005d] in <bcca3bd6208c4f6f9855737db109c2c9>:0 
  at <InteractiveExpressionClass>.Host (System.Object& $retval) [0x00000] in <96f62d92674e4d5d842514b138025ea5>:0 
  at Mono.CSharp.Evaluator.Evaluate (System.String input, System.Object& result, System.Boolean& result_set) [0x00038] in <0b95c870cfeb4d8a87bd60c9d11f7bed>:0 
  at Mono.CSharpShell.Evaluate (System.String input) [0x00000] in <bb84bfc01144492b911515c24eaf53ec>:0 
```

In other words - I prefer to be safe than sorry :)

Reply:
I mis-remembered the docs. :-(

https://docs.microsoft.com/en-us/dotnet/api/system.io.directory.delete?view=netframework-4.7.2

> `DirectoryNotFoundException`: `path` does not exist or could not be found.